network/p2p/throttler.go

```diff
@@ -0,0 +1,48 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"sync"
+
+	"golang.org/x/time/rate"
+
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+var _ Throttler = (*TokenBucketThrottler)(nil)
+
+type Throttler interface {
+	// Throttle blocks until there is [tokens] capacity for [nodeID]
+	Throttle(ctx context.Context, nodeID ids.NodeID, tokens int) error
+}
+
+func NewTokenBucketThrottler(refill rate.Limit, burst int) *TokenBucketThrottler {
```

Comment:
Might be worth linking to https://en.wikipedia.org/wiki/Token_bucket for readers who don't know what a token bucket is